build(aio): ability to generate plunkers

Foxandxss opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

The missing bit here is some extra parameters we give it from the argv. Functionality is there, we just need to find the better way of doing that.

Foxandxss committed:
build(aio): ability to generate plunkers

mary-poppins commented:
The angular.io preview for a4ecac0 is available [here][1].

[1]: https://pr15559-a4ecac0eb6f2fbc3df86c58c4af467da32366872.ngbuilds.io/

gkalpak commented on aio/scripts/plunker-builder/regionExtractor.js:
```diff
@@ -0,0 +1,281 @@
+    case 'ts':
+    case 'js':
+    case 'es6':
+    case 'dart':
```
Is `dart` needed any more?

gkalpak commented on aio/scripts/plunker-builder/regionExtractor.js:
```diff
@@ -0,0 +1,281 @@
+    return  [ { content: content } ];
+  }
+
+  var lines = result = content.split(/\r?\n/);
```
`result` is not used.

gkalpak commented on aio/scripts/plunker-builder/regionExtractor.js:
```diff
@@ -0,0 +1,281 @@
+  if (commentInfo == null) {
+    return content;
+  }
+  var lines = result = content.split(/\r?\n/);
```
`result` is not used.

gkalpak commented on aio/scripts/plunker-builder/regionExtractor.js:
```diff
@@ -0,0 +1,281 @@
+      }
+      if (plasterComment && fragLines.length) {
+        // pad is the padding on the previous line
+        var pad = fragLines[fragLines.length - 1].match(/(\s*)/)[0];
```
Nit: Unnecessary parenthesis around `\s*`.

Foxandxss commented on aio/.gitignore:
```diff
@@ -6,6 +6,9 @@
 /tmp
 
+# generated output
+/content/live-examples
```
Yes, that place is wrong. I need to change it to `src/content/live-examples`

Foxandxss replied:
Fixed.

wardbell commented on aio/scripts/plunker-builder/builder.js:
```diff
@@ -0,0 +1,311 @@
+  _buildCopyrightStrings() {
+    var copyright = 'Copyright 2017 Google Inc. All Rights Reserved.\n'
+      + 'Use of this source code is governed by an MIT-style license that\n'
+      + 'can be found in the LICENSE file at http://angular.io/license';
```
I'm not a lawyer and I don't know what Google's position is. But iirc updating the copyright year is a nicety with no particular legal standing.

Foxandxss replied:
I didn't touch this. We were told to have a licence in the bottom of every file, right?

gkalpak replied:
What I meant was that the text could be changed to `...can be found at http://angular.io/license`.
I assume the current wording is possibly a relic from when the users we pointed to the LICENSE file in the repo. But this is absolutely not worth spending any time on :smiley:

mary-poppins commented:
The angular.io preview for 10e6239 is available [here][1].

[1]: https://pr15559-10e6239f04825a6a1da498412d0cc4bc955fd5b5.ngbuilds.io/

Foxandxss commented on aio/scripts/plunker-builder/builder.js:
```diff
@@ -0,0 +1,311 @@
+    var form = doc.querySelector('form');
+    _.forEach(postData, (value, key) => {
+      var ele = this._htmlToElement(doc, '<input type="hidden" name="' + key + '">');
+      ele.setAttribute('value', value);
```
I don't understand.

gkalpak replied:
Couldn't we create the HTML "by hand" instead of using jsdom?
E.g. something like: 

```js
...
var html = ...
_forEach(postData, (value, key) => {
  html += `<input type="hidden" name="${key}" value="${value}">`;
})
```

Is it so that `value` is properly escaped for use as attribute value?

Foxandxss replied:
I never thought about that George. Someone wrote this before me so I never doubt of the intentions.

That being said, if we can do it without jsdom, I don't see why not.

Foxandxss commented:
I addressed most of your concerns @gkalpak. The `regionExtractor` is now gone. It is part of some transform utils we have.

I left a comment where I had doubts.

On the other hand, I haven't done anything about the translator rules. My regex-fu is not that great but well, what we have work. Feel free to improve them :)

mary-poppins commented:
The angular.io preview for 9a2579c is available [here][1].

[1]: https://pr15559-9a2579cda4261c0be5e7c72c6b4da467cc49d00f.ngbuilds.io/

Foxandxss commented on aio/tools/examples/add-example-boilerplate.js:
```diff
@@ -12,6 +12,7 @@ const EXAMPLES_TESTING_PATH = path.join(EXAMPLES_PATH, 'testing');
 const files = {
   exampleBoilerplate: [
     'src/styles.css',
+    'src/systemjs-angular-loader.js',
```
I added this file "by mistake" but it is a needed fix nonetheless

gkalpak commented on aio/scripts/plunker-builder/translator/fileTranslator.js:
```diff
@@ -0,0 +1,47 @@
+    // rxFrom is a rexexp
+    var rxFrom = rxRule.from;
+    if (rxDatum.from) {
+      var from = rxDatum.from.replace('/', '\/');
```
This still has no effect, but it doesn't break anything either.

gkalpak reviewed (approved):
Since the RegExps work for a while now, I don't feel strongly about improving them (and certainly not as part of this PR). The rest LGTM.

Plnkrs, here we come!
